components/hab/src/command/pkg/uninstall/uninstall_impl.rs

```diff
@@ -330,12 +369,70 @@ fn maybe_delete<U>(ui: &mut U,
         }
         ExecutionStrategy::Run => {
             ui.status(Status::Deleting, &ident)?;
+            if uninstall_hook_mode == UninstallHookMode::Run {
+                maybe_run_uninstall_hook(ui, &install).await?;
+            }
             let pkg_dir = install.installed_path();
             do_clean_delete(&pkg_root_path, &pkg_dir)
         }
     }
 }
 
+/// We only want to run the uninstall hook if this is there are no other revisions.
+/// The uninstall hook is intended to be the inverse of the install hook and is
+/// where one would undo anything performed in the install hook.
+/// Note that install hooks are executed only once when a package is initially
+/// installed and are usually used to perform installation or setup that affects
+/// a machine's global state. For example it might be used to enable a "windows
+/// feature" or invoke a complicated software installer (like sql server). Over
+/// time one might update the package with better logic or adding a software patch.
+/// We only want to run the uninstaller if all revisions are being removed from the
+/// machine. Otherwise if an uninstall hook is run when installing an old version
+/// while a newer version remains installed, that hook could potentially corrupt the
+/// state of the existing package.
+async fn maybe_run_uninstall_hook<T>(ui: &mut T, package: &PackageInstall) -> Result<()>
+    where T: UIWriter
+{
+    let ident = package.ident();
+    let feature_flags = FeatureFlag::from_env(ui);
+    if let Some(ref hook) = UninstallHook::load(&ident.name,
+                                                &svc_hooks_path(ident.name.clone()),
+                                                &package.installed_path.join("hooks"),
+                                                feature_flags)
+    {
+        let unqualified_ident =
+            PackageIdent::from_str(&format!("{}/{}", ident.origin, ident.name))?;
+        let installed = list::package_list(&unqualified_ident.clone().into())?;
+        if installed.len() > 1 {
+            ui.status(Status::Skipping,
+                      format!("execution of uninstall hook for {}. {} packages are installed \
+                               for {}",
+                              ident,
+                              installed.len(),
+                              unqualified_ident))?;
+            return Ok(());
+        }
+
+        ui.status(Status::Executing, format!("uninstall hook for '{}'", ident))?;
+        templating::compile_for_package_install(package, feature_flags).await?;
+        let mut pkg = Pkg::from_install(package).await?;
+        // Only windows uses svc_password
+        if cfg!(target_os = "windows") {
+            // Install hooks do not have access to svc_passwords so
+            // we execute them under the current user account.
+            if let Some(user) = users::get_current_username() {
+                pkg.svc_user = user;
+            }
+        }
```

Comment:
I am sure this ultimately gets compiled into the same thing but we could use a conditional attribute instead of a conditional expression.

```rust
        #[cfg(target_os = "linux")]
        // Install hooks do not have access to svc_passwords so
        // we execute them under the current user account.
        if let Some(user) = users::get_current_username() {
            pkg.svc_user = user;
        }
```